README
Providing a deep view to the CPAN Perl packages
or for any application written in Perl.

WARNING: This is a very early version of the application.
Most of the planned things won't work.

The package contains both the tools to process 
and index the Perl code and the web interface
for it.

It is designed to be able to handle 
- a single perl script 
- a Perl application that lives in a directory tree
- a mirror of CPAN


Platforms:
The application is being developer on Linux and Windows
at the same time so it *should* work on both of them.



SETUP
=======
Mirroring CPAN is done externally.
For example by rsync:
/usr/bin/rsync -av --delete ftp.funet.fi::CPAN /opt/cpan


perl Makefile.PL
make
make test

There is no need to run "make install"


Scripts in the scripts/  directory
=====================================
cpan_digger.pl      - indexing CPAN tree or directory tree


========================================
Index CPAN:
   save the results in the database and ceate static pages
   - unzip file if it is not yet unzipped
   - update database with the following information
     distro:
       name: Dist-Name
       author: <pauseid>
       modules: list of .pm files in lib/  (later also deal with pm files in root)
   - generate html file from pod

Search:
  allow users to search in the database


Server Layout:

/dist/Dist-Name/      <- collected information about the most recent version of the distribution
/dist/Dist-Name/pod/  <- The POD in HTML format
/dist/Dist-Name/src/  <- The source in HTML format with syntax highlighting
    (links to real source lead to the /src/<pauseid>/Dist-Name-1.02/ structure)

/id/<pauseid>/       <- collected information about the author
/id/<pauseid>/Dist-Name-1.02/     <- The POD in HTML format

/src/<pauseid>/Dist-Name-1.02/    <- source code in plain text format (this is where we unzip the files)

/pod/                 <- the documentation of the most recent perl in HTML marked up format
/pod/perl_5_10_01/    <- the documentation of specific versions of perl in HTML marked up formate (Later)

/q                    <- send queries here


robots.txt

set noindex on /id/pauseid/*  (but allow indexing /id/pauseid/ itself)
set noindex on /src/
set noindex on /pod/perl_5_10_01/ and similar


TODO
- allow the user to create all the dependencies of a list of packages and then see "what if I add this extra package?" 
- check why are several versions of the same package processed (eg. Padre)
- List all the licenses that were used in META files and list all the packages with the given license. List all packages without license

- dist page:
  - download link should be based on a configurable cpan server
  - date of release
  - add cpan faces: http://hexten.net/cpan-faces/
  - add links to previous releases (see search.cpan.org)
  - fetch and display number of bugs from RT (and explain why it is not there if the bugtracking is different)
  - Other tools (grep and later diff between versions)
  - fetch data from CPAN Testers and display PASS (146)   NA (1)   UNKNOWN (70)
  - Rating: fetch data from CPANRATINGS and display the stars and the number of reviews
  - For modules, fetch the one-line description and the version number from each one and display them
  - Annotate POD - fetch the number of annotations and display the counter

Be able to index
  1) CPAN or local CPAN with injected files
  2) @INC of current perl
  3) list of directories (e.g. for the @INC of some other perl or file of
     a project in a version control system)

=========================================================
Alternate plan:
*) run CPAN mirror
*) Collect distros:
   Go over all the CPAN mirror directory get all the files.
   For each file adds it to the database AUTHOR, distname, version, path,
   date when we added to the database, date of the file. (see touch how to imitate old dates)
TODO: go over the files reported as error, show them on the web site and think how to include them.

*) Add search engine that can find packages based on their name - fetch always the latest
TODO: show exact hits first
TODO: show newest earlier

TODO: *) Remove distros:
    Go over all the distros in the database and check if they are still in the CPAN mirror.
    If not, mark them as "removed" in the database.
    Maybe also remove them from the extracted directory and
    maybe remove them from the database.
    or just mark them as "removed" in the database?

*) Create a dynamic page for each author with data from the database and
link the results to these pages. -
Generating all the static pages took 75 min on my desktop so we are creating the pages
dynamically for now.

TODO
*) Extract:
    go over all the distros in the database and check if they have been extracted to the 'src' directory already
    if not yet, then try to extract them
    report distros that cannot be extracted (and don't try them again)
*) Generate package/version distro pages:
   Static or dynamic? 
   listing all the versions we have in the database

*) Script that can process an individual distribution which is give as either a zip file
   so we start with unzipping or an already open directory for projects and companies
   that don't package as zip files.

TODO: add a field to each distributions called "processed" which is a timestamp.
Later processes can work on packages that were not yet processed or processedbefore time X.
When displaying a distribution we can know if it was already processed at all and when?
Maybe the processed does not need to be a timestamp but a version number?

1) Run modes:
   - on all CPAN
     fill the author table
     fill the distro table
     unzip distributions
   - on a specific project (not related to CPAN) (or on several projects)
     unzip if needed or copy files
2) In either cases
   - collect data from META files and layout
   - run POD2HTML
   - generate SYN files (PPI)
   - run Perl::Critic   (PPI)
   - Generate Outline   (PPI)
   - MinimumVersion     (PPI)

Later we will add Ajax back to provide hints while the user types in the query.



Changes

0.07 2013.01.13
  - Add all the Perl::Critic::* packages to the list of prereqs.
  - Include the reports from those in the tests.

0.06 2012.06.06
  - fixing some tests
  - stop using YAML::Any

0.05 2012.05.29
  - fixing test

0.04 2012.05.28
  - missing requiremend DBD::SQLite
  - specify the policies better

0.03 2012.05.26
  - replace before by hook before as Dancer has deprecated the former
    (reported by Father Chrysostomos)
  - perltidy the whole project

0.02 2011.05.15
  - Add missing prerequisites.
  - use modules instead of require-ing them.

0.01 2011.05.14
  - Initial version that can already index CPAN and display it
    in an ugly but slightly usable way.


Makefile.PL
use 5.008008;
use strict;
use inc::Module::Install 0.94;

# Workaround for the fact that Module::Install loads the modules
# into memory and when Test::NoWarnings is loaded it will hide
# the warnings generated from that point.
eval {
	require Test::NoWarnings;
	$SIG{__WARN__} = 'DEFAULT';
};

#####################################################################
# Normal Boring Commands

# NOTE: Core modules that aren't dual-life should always have a version of 0

name 'CPAN-Digger';
license 'perl';
author 'Gabor Szabo';
all_from 'lib/CPAN/Digger.pm';

requires 'perl' => '5.008008';

requires 'autodie'       => '2.13';
requires 'Archive::Any'  => '0.0932';
requires 'Archive::Zip'  => '1.30';
requires 'Archive::Tar'  => '1.90';
requires 'Capture::Tiny' => '0.10';
requires 'Carp';
requires 'Cwd';
requires 'Dancer' => '1.3110';
requires 'Data::Dumper';
requires 'DBI'    => '1.623';
requires 'DBD::SQLite' => '1.37';
requires 'Encode' => '2.47';
requires 'File::Basename';
requires 'File::Copy';
requires 'File::Copy::Recursive' => '0.38';
requires 'File::Find::Rule'      => '0.33';
requires 'File::Path';
requires 'File::ShareDir' => '1.03';
requires 'File::Spec';
requires 'File::Temp';
requires 'Getopt::Long';
requires 'JSON::XS' => '2.33';
requires 'JSON'     => '2.53';
requires 'List::Util';
requires 'Moose'                     => '2.0604';
requires 'MooseX::StrictConstructor' => '0.19';

#requires 'Parse::CPAN::Authors';
requires 'Parse::CPAN::Packages'       => '2.38';
requires 'Parse::CPAN::Whois'          => '0.02';

requires 'Perl::Critic'                => '1.118';
requires 'Perl::Critic::Bangs'         => '1.10';
requires 'Perl::Critic::Compatibility' => '1.001';
requires 'Perl::Critic::Deprecated'    => '1.108';
requires 'Perl::Critic::Dynamic'       => '0.05';
requires 'Perl::Critic::Git'           => '1.1.2';
requires 'Perl::Critic::Itch'          => '0.07';
requires 'Perl::Critic::Lax'           => '0.009';
#requires 'Perl::Critic::logicLAB'      => '0.05';
requires 'Perl::Critic::More'          => '1.000';
requires 'Perl::Critic::Nits'          => '1.0.0';
requires 'Perl::Critic::OTRS'          => '0.02';
requires 'Perl::Critic::PetPeeves::JTRAMMELL' => '0.02';
requires 'Perl::Critic::Policy::CodeLayout::TabIndentSpaceAlign'      => '1.0.2';
requires 'Perl::Critic::Policy::Dynamic::NoIndirect'                  => '0.06';
requires 'Perl::Critic::Policy::CodeLayout::ProhibitSpaceIndentation' => '1.02';
requires 'Perl::Critic::Policy::logicLAB::RequireSheBang'             => '0.03';
requires 'Perl::Critic::Policy::logicLAB::RequireVersionFormat'       => '0.03';
requires 'Perl::Critic::Policy::logicLAB::ProhibitUseLib'             => '0.01';
requires 'Perl::Critic::Policy::logicLAB::ProhibitShellDispatch'      => '0.01';
requires 'Perl::Critic::RENEEB'          => '0.01';
requires 'Perl::Critic::StricterSubs'    => '0.03';
requires 'Perl::Critic::Storable'        => '0.01';
requires 'Perl::Critic::Swift'           => '1.0.3';
requires 'Perl::Critic::Pulp'            => '75';
requires 'Perl::Critic::Tics'            => '0.006';

requires 'Perl::MinimumVersion'          => '1.31';
requires 'POSIX';
requires 'PPI::Document'                 => '1.215'; # PPI
requires 'PPI::Find'                     => '1.215'; # PPI
requires 'PPIx::EditorTools'             => '0.18';
requires 'Template'                      => '2.24';
requires 'Time::HiRes';
#requires 'YAML::Any' => '0.72';
requires 'YAML'      => '0.84';

requires 'Pod::Simple' => '3.23';
requires 'Pod::Simple::HTML';  # in Pod-Simple
requires 'Plack' => '1.0014';
requires 'Plack::Runner'; # no version ? in Plack
requires 'Plack::Handler::FCGI'; # in Plack
requires 'Plack::Util';  # in Plack


test_requires 'Test::Deep'       => '0.110';
test_requires 'Test::More'       => '0.98';
test_requires 'Test::NoWarnings' => '1.04';

no_index 'directory' => qw{ t xt eg share inc privinc };

install_script 'script/cpan_digger.pl';
homepage 'http://cpandigger.com/';
repository 'https://github.com/szabgab/CPAN-Digger';

eval {
	require File::Copy::Recursive;
	mkdir 'share';

	# config.yml
	# environmaents
	foreach my $dir (qw(public schema bin view)) {
		File::Copy::Recursive::rcopy( $dir, "share/$dir" );
	}

	install_share();
};

WriteAll;

